File: modules/domain/src/main/scala/com/streamfirst/iceberg/hybrid/domain/TableMetadata.scala
package com.streamfirst.iceberg.hybrid.domain

import java.time.Instant

/** Represents a specific version of Apache Iceberg table metadata. Contains all information needed
  * to describe the table's structure and data files at a particular point in time. Metadata is
  * versioned by commit ID and replicated across regions for geo-distributed access.
  */
final case class TableMetadata(
  tableId: TableId,
  commitId: CommitId,
  sourceRegion: Region,
  timestamp: Instant,
  dataFiles: List[StoragePath],
  schema: String
):
  override def toString: String =
    s"TableMetadata(tableId=$tableId, commitId=$commitId, sourceRegion=${sourceRegion.id}, " +
      s"timestamp=$timestamp, dataFileCount=${dataFiles.size})"

object TableMetadata:
  def create(
    tableId: TableId,
    sourceRegion: Region,
    dataFiles: List[StoragePath],
    schema: String
  ): TableMetadata =
    TableMetadata(
      tableId = tableId,
      commitId = CommitId.generate(),
      sourceRegion = sourceRegion,
      timestamp = Instant.now(),
      dataFiles = dataFiles,
      schema = schema
    )

File: legacy-java/domain-java/src/main/java/com/streamfirst/iceberg/hybrid/domain/TableMetadata.java
package com.streamfirst.iceberg.hybrid.domain;

import java.time.Instant;
import java.util.List;
import lombok.EqualsAndHashCode;
import lombok.NonNull;
import lombok.Value;

/**
 * Represents a specific version of Apache Iceberg table metadata. Contains all information needed
 * to describe the table's structure and data files at a particular point in time. Metadata is
 * versioned by commit ID and replicated across regions for geo-distributed access.
 */
@Value
@EqualsAndHashCode(of = {"tableId", "commitId"})
public class TableMetadata {
  /** The table this metadata describes */
  @NonNull TableId tableId;

  /** Unique identifier for this metadata version */
  @NonNull CommitId commitId;

  /** The region where this metadata version was originally created */
  @NonNull Region sourceRegion;

  /** When this metadata version was created */
  @NonNull Instant timestamp;

  /** List of data file paths that comprise the table at this version */
  @NonNull List<StoragePath> dataFiles;

  /** The table schema as JSON (Iceberg schema format) */
  @NonNull String schema;

  @Override
  public String toString() {
    return "TableMetadata{"
        + "tableId="
        + tableId
        + ", commitId="
        + commitId
        + ", sourceRegion="
        + sourceRegion
        + ", timestamp="
        + timestamp
        + ", dataFileCount="
        + dataFiles.size()
        + '}';
  }
}
